From 4e1f2c08dbe28bb25950fed423bf4184aaa555a3 Mon Sep 17 00:00:00 2001
From: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Date: Wed, 17 Apr 2024 11:49:03 +0200
Subject: [PATCH] rpmsg: i2c: fix structure initialization leading to memory
 leak

I2C rpmsg structure needs to have at least 2 fields and the last one
must be empty. It is used in rpmsg_dev_match function as criteria
to stop parsing rpmsg driver name by detecting empty field name.

[LOG]
[   27.740602] ==================================================================
[   27.746808] BUG: KASAN: slab-out-of-bounds in rpmsg_dev_match+0xa4/0x140
[   27.753435] Read of size 1 at addr ffff00008b64e5e8 by task irq/18-40490000/95
[   27.760751]
[   27.762202] CPU: 0 PID: 95 Comm: irq/18-40490000 Tainted: G           O       6.1.73 #1
[   27.770215] Hardware name: STMicroelectronics STM32MP257F-EV1 Evaluation Board (DT)
[   27.777813] Call trace:
[   27.780265]  dump_backtrace+0xdc/0x130
[   27.783962]  show_stack+0x18/0x30
[   27.787340]  dump_stack_lvl+0x64/0x80
[   27.790929]  print_report+0x154/0x458
[   27.794622]  kasan_report+0xb8/0x100
[   27.798213]  __asan_load1+0x68/0x74
[   27.801692]  rpmsg_dev_match+0xa4/0x140
[   27.805479]  __device_attach_driver+0x50/0x1a0
[   27.884652]
[   27.886093] Allocated by task 497:
[   27.889553]  kasan_save_stack+0x3c/0x70
[   27.893340]  kasan_set_track+0x2c/0x40
[   27.897119]  kasan_save_alloc_info+0x24/0x34
[   27.901409]  __kasan_kmalloc+0xb8/0xc0
[   27.905093]  __kmalloc_node_track_caller+0x60/0xa0
[   27.909882]  devm_kmalloc+0x54/0x12c
[   27.913465]  rpmsg_i2c_platform_probe+0x9c/0x290 [i2c_rpmsg]
[   27.919123]  platform_probe+0x8c/0xf0

Change-Id: I6ff906ccdbab10d0e2314fb496d35166d975db15
Signed-off-by: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/374730
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
---
 drivers/i2c/busses/i2c-rpmsg.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/i2c/busses/i2c-rpmsg.c b/drivers/i2c/busses/i2c-rpmsg.c
index aea8ce069597..fa0514da97bd 100644
--- a/drivers/i2c/busses/i2c-rpmsg.c
+++ b/drivers/i2c/busses/i2c-rpmsg.c
@@ -65,7 +65,7 @@ struct rpmsg_i2c_dev {
 
 	struct list_head list; /* adapter device list */
 	struct rpmsg_driver rpdrv;
-	struct rpmsg_device_id dev_id;
+	struct rpmsg_device_id dev_id[2];
 	struct device *dev;
 
 	struct rpmsg_device *rpdev;
@@ -258,7 +258,7 @@ static int rpmsg_i2c_probe(struct rpmsg_device *rpdev)
 	mutex_lock(&rpmsg_i2c_lock);
 
 	list_for_each_entry(ri2c_dev, &rpmsg_i2c_list, list) {
-		if (!strcmp(ri2c_dev->dev_id.name, rpdev->id.name)) {
+		if (!strcmp(ri2c_dev->dev_id[0].name, rpdev->id.name)) {
 			if (ri2c_dev->rpdev) {
 				ret = -EBUSY;
 				goto out;
@@ -356,10 +356,10 @@ static int rpmsg_i2c_platform_probe(struct platform_device *pdev)
 		dev_err(dev, "Error, proc-id property is missing: %d\n", ret);
 		return ret;
 	}
-	strncpy(ri2c_dev->dev_id.name, dev_id_name, RPMSG_NAME_SIZE - 1);
+	strscpy_pad(ri2c_dev->dev_id[0].name, dev_id_name, RPMSG_NAME_SIZE);
 
 	ri2c_dev->rpdrv.drv.name = "rpmsg_i2c";
-	ri2c_dev->rpdrv.id_table = &ri2c_dev->dev_id;
+	ri2c_dev->rpdrv.id_table = ri2c_dev->dev_id;
 	ri2c_dev->rpdrv.probe = rpmsg_i2c_probe;
 	ri2c_dev->rpdrv.callback = rpmsg_i2c_cb;
 	ri2c_dev->rpdrv.remove = rpmsg_i2c_remove;
-- 
2.39.5

